# Detect Phishing URLs
### Capstone 3 - Data Wrangling and EDA
Michael Garber

#### High-Level Steps
1. Data Wrangling
    1. Data Collection
    2. Data Organization
    3. Data Definition
    4. Data Cleaning
1. Exploratory Data Analysis


#### Data Wrangling

##### Data Collection
URL Data
- https://www.kaggle.com/datasets/hassaanmustafavi/phishing-urls-dataset

TLD (Top-level Domain) Data
- https://www.iana.org/domains/root/db

In [6]:
# Please Install KaggleHub if you don't already have it
%pip install kagglehub

In [7]:
# Please Installtldextract if you don't already have it
%pip install tldextract

In [8]:
# Import Packages
import pandas as pd
import kagglehub
import os
import shutil
from urllib.parse import urlparse
import tldextract
import ipaddress

In [9]:
# Download Data
kPath = kagglehub.dataset_download("hassaanmustafavi/phishing-urls-dataset")

In [10]:
# build source file path
fileName = 'url_dataset.csv'
srcFilePath = os.path.join(kPath, fileName)
dataDir = os.path.join(os.pardir, 'data', 'raw', fileName)

# Move File to project folder
shutil.copy(srcFilePath, dataDir)

'..\\data\\raw\\url_dataset.csv'

In [11]:
# Create pandas dataframe - urlData
urlData = pd.read_csv(dataDir)

In [12]:
# Create pandas dataframe - TLDs (Top-level Domains)
tldDir = '../data/raw/TLDs.csv'
tldNames = pd.read_csv(tldDir)

# TLDs source...
# https://www.iana.org/domains/root/db

In [13]:
# check dataframe urlData - # or rows
print(urlData.shape)

(450176, 2)


In [14]:
# check dataframe tldNames - # or rows
print(tldNames.shape)

(1591, 3)


##### Data Organization
Project file structure based on the cookiecutter data science template. \
[https://drivendata.github.io/cookiecutter-data-science/](https://drivendata.github.io/cookiecutter-data-science/)

Folder structure tree (GitHub) \
[https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md](https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md)

##### Data Definition

- Column names
- Data types
- Description of the columns
- Counts and unique values
- Ranges of values
- Calc Summary statistics

In [17]:
# Data types, unique values, range of index
urlData.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450176 entries, 0 to 450175
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   url     450176 non-null  object
 1   type    450176 non-null  object
dtypes: object(2)
memory usage: 6.9+ MB


Description of the columns
- url - The web address to be analyzed.
- type - The classification of the URL (phishing or legitimate).

In [19]:
# increase pandas dataframe column width - full size
pd.set_option('display.width', None)

In [20]:
# Value counts for URLs
pd.DataFrame(urlData['url' ].value_counts())

count
url                                                      
https://www.google.com                                  1
https://www.tabheaven.com/gales-eric-tabs.html          1
https://www.tabs-database.com/justin-mcroberts-...      1
https://www.tabpower.com/a806.html                      1
https://www.tabor.edu/alumni/directory?decade=1960      1
...                                                   ...
https://www.billboard.com/artist/anita-pointer/...      1
https://www.billboard.com/artist/andy-kim/chart...      1
https://www.billboard.com/artist/alvino-rey/dis...      1
https://www.billboard.com/artist/3x-krazy/disco...      1
http://writeassociate.com/test/Portal/inicio/IO...      1

[450176 rows x 1 columns]

- These URLs above are all unique ^

In [22]:
pd.DataFrame(urlData['type' ].value_counts()).head()

count
type              
legitimate  345738
phishing    104438

In [23]:
# describe dataframe
urlData.describe()

url        type
count                   450176      450176
unique                  450176           2
top     https://www.google.com  legitimate
freq                         1      345738

Data Definition Summary

- Column names
    - **url**
        - Data types
            - series of string
        - Description of the column 
            - The web address to be analyzed.
        - Counts and unique values 
            - 450176 (all unique)
        - Ranges of values 
            - all unique URLs
    - **type**
        - Data types 
            - series of string
        - Description of the column
            - The classification of the URL (phishing or legitimate).
        - Counts and unique values
            - legitimate    345738
            - phishing      104438
        - Ranges of values
            - phishing
            - legitimate
            
- Calc Summary statistics
    - N/A ... text data

##### Data Cleaning

- this dataset is mostly clean  =)
- identify invalid URLs (those that urllib couldn't parse) and remove

#### Exploratory Data Analysis

- create new features from parsed URLs
- do value counts on the new url parts
- check correlations of the url parts to the target variable

In [27]:
# view the data head - URLs
urlData.head(10)

url        type
0     https://www.google.com  legitimate
1    https://www.youtube.com  legitimate
2   https://www.facebook.com  legitimate
3      https://www.baidu.com  legitimate
4  https://www.wikipedia.org  legitimate
5     https://www.reddit.com  legitimate
6      https://www.yahoo.com  legitimate
7   https://www.google.co.in  legitimate
8         https://www.qq.com  legitimate
9     https://www.amazon.com  legitimate

In [28]:
# view the data head - only phish rows
urlData[urlData['type'] == 'phishing'].head(10)

url      type
345738                   http://atualizacaodedados.online  phishing
345739                     http://webmasteradmin.ukit.me/  phishing
345740           http://stcdxmt.bigperl.in/klxtv/apps/uk/  phishing
345741   https://tubuh-syarikat.com/plugins/fields/files/  phishing
345742  http://rolyborgesmd.com/exceword/excel.php?.ra...  phishing
345743  http://ongelezen-voda.000webhostapp.com/inlogg...  phishing
345744  http://www.valenzaceramic.com/home/webapps/e52...  phishing
345745  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345746  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345747  http://chronopost-service-enligne.net/56123s/r...  phishing

##### TLD Data
> "Top level domains" ...to be integrated with url data set

In [30]:
# view the data (to be used as metadata for URLs)
tldNames

Domain          Type  \
0         .aaa       generic   
1        .aarp       generic   
2      .abarth       generic   
3         .abb       generic   
4      .abbott       generic   
...        ...           ...   
1586    .zippo       generic   
1587       .zm  country-code   
1588     .zone       generic   
1589  .zuerich       generic   
1590       .zw  country-code   

                                            TLD Manager  
0                 American Automobile Association, Inc.  
1                                                  AARP  
2                                          Not assigned  
3                                               ABB Ltd  
4                             Abbott Laboratories, Inc.  
...                                                 ...  
1586                                       Not assigned  
1587  Zambia Information and Communications Technolo...  
1588                                    Binky Moon, LLC  
1589                   Kanton Zürich (Canton of Zurich)  
1590  Postal and Telecommunications Regulatory Autho...  

[1591 rows x 3 columns]

In [31]:
# describe tldNames
tldNames.describe()

Domain     Type      TLD Manager
count    1591     1591             1591
unique   1591        6              754
top      .aaa  generic  Binky Moon, LLC
freq        1     1246              196

In [32]:
# get the different TLD types
pd.DataFrame(tldNames['Type'].value_counts())

count
Type                     
generic              1246
country-code          316
sponsored              14
test                   11
generic-restricted      3
infrastructure          1

> **Country code** appears to be the most meaningful type among TLD type

##### URL Parsing

In [35]:
# create function urlparse_try 
#    otherwise, parsing will fail on an invalid URL
def urlparse_try(url):
    try:
        return urlparse(url)
    except:
        return 'invalid URL'

In [36]:
# engineer new feature for EDA - parsedUrl
urlData['parsedUrl'] = urlData['url'].apply(urlparse_try)

In [37]:
# example of a parsed url
print('Example of a parsed url \n')
print(urlparse(urlData['url'][1]))

Example of a parsed url 

ParseResult(scheme='https', netloc='www.youtube.com', path='', params='', query='', fragment='')


Parsed URL parts
> scheme://netloc/path;parameters?query#fragment

> **URL Part Reference**

![Image](../references/URL_part_diagram.png)

- image from https://www.geeksforgeeks.org/components-of-a-url/


In [40]:
# find invalid URLs (urls that failed urllib parsing...those will cause errors during processing)
invalidUrlRows = urlData[urlData['parsedUrl'] == 'invalid URL']
pd.DataFrame(invalidUrlRows).head()

url      type  \
397556  http://ladiesfirst-privileges[.]com/656465/d56...  phishing   

          parsedUrl  
397556  invalid URL

In [41]:
# remove invalid URLs (they will causes errors)
urlData = urlData.drop(invalidUrlRows.index)

In [42]:
# confirm 1 record removed
urlData.shape

(450175, 3)

Engineer new features from parsedUrl

In [44]:
# scheme
urlData['urlPart_scheme'] = urlData['parsedUrl'].apply(lambda urlObj : urlObj.scheme)

In [45]:
# create feature - subdomain
urlData['subDomain'] = urlData['url'].apply(lambda url : tldextract.extract(url).subdomain)

# create feature - subdomain
urlData['domain'] = urlData['url'].apply(lambda url : tldextract.extract(url).domain)

# create feature - tld (top-level domain)
urlData['tld'] = urlData['url'].apply(lambda url : tldextract.extract(url).suffix)

In [46]:
# make try except for port because it fails when port # is NULL
# def retPort():
#     try:
#         urlObj.port
#     except:
#         None

In [47]:
# port - ***no port numbers in the data so removing this code***
# urlData['urlPart_port'] = urlData['parsedUrl'].apply(lambda urlObj : retPort())

#  create feature - path
urlData['urlPart_path'] = urlData['parsedUrl'].apply(lambda urlObj : urlObj.path)

# create feature - query
urlData['urlPart_query'] = urlData['parsedUrl'].apply(lambda urlObj : urlObj.query)

# create feature - fragement
urlData['urlPart_fragment'] = urlData['parsedUrl'].apply(lambda urlObj : urlObj.fragment)

In [48]:
urlData.head(1)

url        type                        parsedUrl  \
0  https://www.google.com  legitimate  (https, www.google.com, , , , )   

  urlPart_scheme subDomain  domain  tld urlPart_path urlPart_query  \
0          https       www  google  com                              

  urlPart_fragment  
0

In [49]:
tldNames.head(1)

Domain     Type                            TLD Manager
0   .aaa  generic  American Automobile Association, Inc.

In [50]:
# remove initial period from tlds (as that will prevent merge/joining)
tldsNoPeriod = tldNames['Domain'].apply(lambda tld : tld[1:] if tld[0] == '.' else tld)

In [51]:
urlData['tld'].head(30)

0        com
1        com
2        com
3        com
4        org
5        com
6        com
7      co.in
8        com
9        com
10       com
11       com
12       com
13     co.jp
14       com
15       com
16       com
17       com
18    com.cn
19       com
20       com
21        cn
22        de
23     co.uk
24    com.br
25        fr
26        ru
27       com
28        ru
29       com
Name: tld, dtype: object

In [52]:
# create column for urlData.tld with period and part before period removed
urlData['tld_join'] = urlData['tld'].apply(lambda tld : tld[(tld.find('.') + 1):] if tld.find('.') != -1 else tld)

In [53]:
# add rows from TLD data set
urlData = pd.merge(urlData, tldNames, left_on=urlData.tld_join, right_on=tldsNoPeriod, how='left')

In [54]:
urlData.head(30)

key_0                         url        type  \
0    com      https://www.google.com  legitimate   
1    com     https://www.youtube.com  legitimate   
2    com    https://www.facebook.com  legitimate   
3    com       https://www.baidu.com  legitimate   
4    org   https://www.wikipedia.org  legitimate   
5    com      https://www.reddit.com  legitimate   
6    com       https://www.yahoo.com  legitimate   
7     in    https://www.google.co.in  legitimate   
8    com          https://www.qq.com  legitimate   
9    com      https://www.amazon.com  legitimate   
10   com      https://www.taobao.com  legitimate   
11   com     https://www.twitter.com  legitimate   
12   com       https://www.tmall.com  legitimate   
13    jp    https://www.google.co.jp  legitimate   
14   com          https://www.vk.com  legitimate   
15   com        https://www.live.com  legitimate   
16   com   https://www.instagram.com  legitimate   
17   com        https://www.sohu.com  legitimate   
18    cn     https://www.sina.com.cn  legitimate   
19   com          https://www.jd.com  legitimate   
20   com       https://www.weibo.com  legitimate   
21    cn          https://www.360.cn  legitimate   
22    de       https://www.google.de  legitimate   
23    uk    https://www.google.co.uk  legitimate   
24    br   https://www.google.com.br  legitimate   
25    fr       https://www.google.fr  legitimate   
26    ru       https://www.google.ru  legitimate   
27   com     https://www.netflix.com  legitimate   
28    ru       https://www.yandex.ru  legitimate   
29   com  https://www.list.tmall.com  legitimate   

                              parsedUrl urlPart_scheme subDomain     domain  \
0       (https, www.google.com, , , , )          https       www     google   
1      (https, www.youtube.com, , , , )          https       www    youtube   
2     (https, www.facebook.com, , , , )          https       www   facebook   
3        (https, www.baidu.com, , , , )          https       www      baidu   
4    (https, www.wikipedia.org, , , , )          https       www  wikipedia   
5       (https, www.reddit.com, , , , )          https       www     reddit   
6        (https, www.yahoo.com, , , , )          https       www      yahoo   
7     (https, www.google.co.in, , , , )          https       www     google   
8           (https, www.qq.com, , , , )          https       www         qq   
9       (https, www.amazon.com, , , , )          https       www     amazon   
10      (https, www.taobao.com, , , , )          https       www     taobao   
11     (https, www.twitter.com, , , , )          https       www    twitter   
12       (https, www.tmall.com, , , , )          https       www      tmall   
13    (https, www.google.co.jp, , , , )          https       www     google   
14          (https, www.vk.com, , , , )          https       www         vk   
15        (https, www.live.com, , , , )          https       www       live   
16   (https, www.instagram.com, , , , )          https       www  instagram   
17        (https, www.sohu.com, , , , )          https       www       sohu   
18     (https, www.sina.com.cn, , , , )          https       www       sina   
19          (https, www.jd.com, , , , )          https       www         jd   
20       (https, www.weibo.com, , , , )          https       www      weibo   
21          (https, www.360.cn, , , , )          https       www        360   
22       (https, www.google.de, , , , )          https       www     google   
23    (https, www.google.co.uk, , , , )          https       www     google   
24   (https, www.google.com.br, , , , )          https       www     google   
25       (https, www.google.fr, , , , )          https       www     google   
26       (https, www.google.ru, , , , )          https       www     google   
27     (https, www.netflix.com, , , , )          https       www    netflix   
28       (https, www.yandex.ru, , , , )          https       www     yandex   
29  (https, www.list

##### Is IP Address?

In [56]:
# check if netloc is an IP address function
def is_ip_address(netloc):
    try:
        ipaddress.ip_address(netloc)
        return True
    except ValueError:
        return False

In [57]:
# create feature - isIPaddress
urlData['isIPaddress'] = urlData['parsedUrl'].apply(lambda urlObj : is_ip_address(urlObj.netloc))

In [58]:
# new features - check head
urlData.head()

key_0                        url        type  \
0   com     https://www.google.com  legitimate   
1   com    https://www.youtube.com  legitimate   
2   com   https://www.facebook.com  legitimate   
3   com      https://www.baidu.com  legitimate   
4   org  https://www.wikipedia.org  legitimate   

                            parsedUrl urlPart_scheme subDomain     domain  \
0     (https, www.google.com, , , , )          https       www     google   
1    (https, www.youtube.com, , , , )          https       www    youtube   
2   (https, www.facebook.com, , , , )          https       www   facebook   
3      (https, www.baidu.com, , , , )          https       www      baidu   
4  (https, www.wikipedia.org, , , , )          https       www  wikipedia   

   tld urlPart_path urlPart_query urlPart_fragment tld_join Domain     Type  \
0  com                                                  com   .com  generic   
1  com                                                  com   .com  generic   
2  com                                                  com   .com  generic   
3  com                                                  com   .com  generic   
4  org                                                  org   .org  generic   

                         TLD Manager  isIPaddress  
0  VeriSign Global Registry Services        False  
1  VeriSign Global Registry Services        False  
2  VeriSign Global Registry Services        False  
3  VeriSign Global Registry Services        False  
4     Public Interest Registry (PIR)        False

##### Correlation to Target

In [60]:
# indices of all phish urls
isPhishIdx = urlData[urlData['type'] == 'phishing'].index

In [61]:
# Top phishing schemes
pd.DataFrame(urlData[['urlPart_scheme', 'type']].loc[isPhishIdx].value_counts()).head()

count
urlPart_scheme type           
http           phishing  97947
https          phishing   6483
ftp            phishing      7

In [62]:
# Top phishing domains
pd.DataFrame(urlData[['domain', 'type']].loc[isPhishIdx].value_counts()).head()

count
domain        type           
google        phishing   1325
000webhostapp phishing    978
mhcable       phishing    381
kf25zx        phishing    265
adnxs         phishing    255

In [63]:
# Top phishing subdomains
pd.DataFrame(urlData[['subDomain', 'type']].loc[isPhishIdx].value_counts()).head()

count
subDomain type           
          phishing  69955
www       phishing  13014
sites     phishing    982
webmail   phishing    397
mail      phishing    177

In [64]:
# Top phishing TLDs
pd.DataFrame(urlData[['tld', 'type']].loc[isPhishIdx].value_counts()).head(10)

count
tld    type           
com    phishing  51092
net    phishing   5242
       phishing   4321
org    phishing   3540
ru     phishing   3523
com.br phishing   3319
info   phishing   1827
com.au phishing   1822
pl     phishing   1326
co.uk  phishing   1295

In [65]:
# % of IP address URLs that is phish
pd.crosstab(urlData['type'], urlData['isIPaddress'])

isIPaddress   False  True 
type                      
legitimate   345738      0
phishing     101571   2866

> 100% of the netlocs that were in IP address format URLs were phishing in this data set

In [67]:
# add a new column 'isPhish_bool' of bools for numerical analysis of 'type' column
urlData['isPhish_bool'] = urlData['type'].apply(lambda type : True if type == 'phishing' else False)

In [68]:
urlData.head(1)

key_0                     url        type                        parsedUrl  \
0   com  https://www.google.com  legitimate  (https, www.google.com, , , , )   

  urlPart_scheme subDomain  domain  tld urlPart_path urlPart_query  \
0          https       www  google  com                              

  urlPart_fragment tld_join Domain     Type  \
0                       com   .com  generic   

                         TLD Manager  isIPaddress  isPhish_bool  
0  VeriSign Global Registry Services        False         False

In [69]:
# pivot table - tld vs target
pd.pivot_table(urlData, index=['tld_join'], values='isPhish_bool', aggfunc=['count', 'mean']).sort_values(by=[('mean', 'isPhish_bool'), ('count', 'isPhish_bool')], ascending=False).head(30)

count         mean
         isPhish_bool isPhish_bool
tld_join                          
                 4321          1.0
top               513          1.0
xyz               425          1.0
ga                235          1.0
gq                186          1.0
ml                144          1.0
cf                134          1.0
work              132          1.0
click             100          1.0
ms                 90          1.0
website            80          1.0
nf                 70          1.0
bo                 65          1.0
onion              41          1.0
xn--p1ai           37          1.0
site               36          1.0
mz                 35          1.0
COM                32          1.0
tz                 31          1.0
gd                 29          1.0
gdn                29          1.0
link               28          1.0
space              22          1.0
wang               21          1.0
host               20          1.0
legal              19          1.0
cr                 17          1.0
party              16          1.0
re                 12          1.0
live               11          1.0

In [70]:
# pivot table - Top Country-specific domains used by phishers
pd.pivot_table(urlData[urlData['Type'] == 'country-code'], index=['tld_join', 'TLD Manager'], values='isPhish_bool', aggfunc=['count', 'mean']).sort_values(by=[('mean', 'isPhish_bool'), ('count', 'isPhish_bool')], ascending=False)

count  \
                                                            isPhish_bool   
tld_join TLD Manager                                                       
ga       Agence Nationale des Infrastructures Numériques...          235   
gq       GETESA                                                      186   
ml       Agence des Technologies de l’Information et de ...          144   
cf       Societe Centrafricaine de Telecommunications (S...          134   
ms       MNI Networks Ltd.                                            90   
...                                                                  ...   
bj       Autorité de Régulation des Communications Elect...            1   
iq       Communications and Media Commission (CMC)                     1   
kw       Communications and Information Technology Regul...            1   
mc       Direction des Plateformes et des Ressources Num...            1   
pr       Gauss Research Laboratory Inc.                                1   

                                                                    mean  
                                                            isPhish_bool  
tld_join TLD Manager                                                      
ga       Agence Nationale des Infrastructures Numériques...          1.0  
gq       GETESA                                                      1.0  
ml       Agence des Technologies de l’Information et de ...          1.0  
cf       Societe Centrafricaine de Telecommunications (S...          1.0  
ms       MNI Networks Ltd.                                           1.0  
...                                                                  ...  
bj       Autorité de Régulation des Communications Elect...          0.0  
iq       Communications and Media Commission (CMC)                   0.0  
kw       Communications and Information Technology Regul...          0.0  
mc       Direction des Plateformes et des Ressources Num...          0.0  
pr       Gauss Research Laboratory Inc.                              0.0  

[205 rows x 2 columns]

In [71]:
# Export df to file
urlData.to_csv('../data/interim/urlData_raw.csv')